### Python numbers

In [16]:
2 + 1 # plus

3

In [2]:
2 - 1 # minus

1

In [3]:
2 * 2 # multiply

4

In [4]:
2 / 3 # divide

0.6666666666666666

In [6]:
7 % 4 # Modulo

3

In [9]:
(-7) % 3 # Modulo

2

In [10]:
2 ** 3 # power

8

`https://docs.python.org/2/tutorial/floatingpoint.html`
No matter how many base 2 digits you’re willing to use, the decimal value 0.1 cannot be represented exactly as a base 2 fraction. In base 2, 1/10 is the infinitely repeating fraction. On a typical machine running Python, there are 53 bits of precision available for a Python float. 

It’s easy to forget that the stored value is an approximation to the original decimal fraction, because of the way that floats are displayed at the interpreter prompt.

In [12]:
0.1+0.2 # approximation with decimal

0.30000000000000004

Other surprises follow from this one. 
For example, if you try to round the value 2.675 to two decimal places, you get this:

In [14]:
round(2.675, 2) # surprise

2.67

Since the decimal fraction 2.675 is exactly halfway between 2.67 and 2.68, you might expect the result here to be (a binary approximation to) 2.68. It’s not, because when the decimal string 2.675 is converted to a binary floating-point number, it’s again replaced with a binary approximation, whose exact value is 2.67499999999999982236431605997495353221893310546875. 

If you’re in a situation where you care which way your decimal halfway-cases are rounded, you should consider using the decimal module, which provides a nice way to “see” the exact value that’s stored in any particular Python float:

In [15]:
from decimal import Decimal
Decimal(2.675) # to see the exact value 

Decimal('2.67499999999999982236431605997495353221893310546875')

Still, don’t be unduly wary of floating-point! The errors in Python float operations are inherited from the floating-point hardware, and on most machines are on the order of no more than 1 part in `2**53` per operation. That’s more than adequate for most tasks, but you do need to keep in mind that it’s not decimal arithmetic, and that every float operation can suffer a new rounding error.

While pathological cases do exist, for most casual use of floating-point arithmetic you’ll see the result you expect in the end if you simply round the display of your final results to the number of decimal digits you expect.

### Variable Assignments

In [17]:
a = 5

In [18]:
a

5

In [19]:
a = a + a

In [20]:
a

10

In [21]:
type(a) # return type of a variable

int

In [22]:
a = 3.2 # example of dynamic typing

In [24]:
type(a)

float

In [25]:
income = 3
tax_rate = 0.2
tax_amount = income * tax_rate

In [26]:
tax_amount

0.6000000000000001

### Strings

In [1]:
'hello' # string example

'hello'

In [3]:
'This is a string'

'This is a string'

In [4]:
"I'm learning Python" # take advantage of single or double quotes

"I'm learning Python"

In [6]:
print('hello')
print("world")

hello
world


In [7]:
print('hello \n world') # have a new line in a string

hello 
 world


In [8]:
print('hello \tworld') # have a tab in a string

hello 	world


In [10]:
len('hello round') # length of string

11

### Indexing and Slicing with Strings

In [1]:
s = "Hello World!"

In [2]:
s[0] # get a char out of string

'H'

In [4]:
s[9]

'l'

In [8]:
s[-3] # reverese indexing

'l'

In [9]:
s[2:] # get substr starting at idx=2

'llo World!'

In [10]:
s[:2] # get the substr up to but exclusive idx=2

'He'

In [11]:
s[2:5] # get the substr start at idx=2 and ends at idx=5 (exclusive)

'llo'

In [12]:
s[::] # get the whole string, with stepsize unspecified

'Hello World!'

In [13]:
s[::2] # get the whole string with stepsize=2

'HloWrd'

In [14]:
s[2:7:2] # get the substring idx = [2, 7) with stepsize=2. [start:stop:stepsize]

'loW'

In [15]:
s[::-1] # reverse the string

'!dlroW olleH'

### String properties and methods

In [27]:
name = 'Sam'

In [18]:
name[0] = 'P' # TypeError due to string is immutable
# So if want to change it, will need to create a new string. 

TypeError: 'str' object does not support item assignment

In [28]:
name2 = 'P' + name[1:] # string concatenation

In [29]:
name2

'Pam'

In [30]:
name3 = name * 5 # multiplication of strings

In [31]:
name3

'SamSamSamSamSam'

In [32]:
# type name. then press tab key would show all string methods
name.upper() # convert to upper case
# this do not affect the name string
# you need to assign it to save the result to a variable

'SAM'

In [33]:
"NAME".lower() # convert to lower case

'name'

In [34]:
"Hello World!".split() # split a string by space, return a list

['Hello', 'World!']

In [35]:
"Hello World!".split("o") # split by specified string

['Hell', ' W', 'rld!']

In [36]:
"Hello World!".split("o ") # split by specified string

['Hell', 'World!']

### Print formatting with strings

String interpolation (stick a var into a string)

In [37]:
print('This is a string {}'.format('INSERTED'))

This is a string INSERTED


In [39]:
print('The {} {} {}'.format('round1', 'round2', 'round3'))

The round1 round2 round3


In [40]:
print('The {2} {1} {0}'.format('round1', 'round2', 'round3')) # reorder

The round3 round2 round1


In [41]:
print('The {0} {0} {0}'.format('round1', 'round2', 'round3')) # repeat

The round1 round1 round1


In [43]:
print('The {r1} {r2} {r3}'.format(r1='round1', r2='round2', r3='round3')) # repeat# assign keywords

The round1 round2 round3


In [44]:
result = 100/777 # floating point formatting

In [45]:
result

0.1287001287001287

In [46]:
print("The result was {}".format(result))

The result was 0.1287001287001287


In [47]:
print("The result was {r}".format(r=result))

The result was 0.1287001287001287


In [48]:
print("The result was {r:1.3f}".format(r=result)) # {value[:width][.precision]f}

The result was 0.129


In [49]:
print("The result was {r:10.3f}".format(r=result))

The result was      0.129


In [50]:
print("The result was {r:0.3f}".format(r=result))

The result was 0.129


In [55]:
print("The result was {r:2.5f}".format(r=result))

The result was 0.12870


In [56]:
result = 104.12345

In [57]:
print("The result was {r:1.2f}".format(r=result))

The result was 104.12


In [58]:
name = 'Jose'

In [59]:
print(f'Hello, his name is {name}') # introduced in Python 3.6

Hello, his name is Jose


In [60]:
name = "Sam"
age = 3

In [61]:
print(f"{name}'s age is {age}")

Sam's age is 3


## Lists

In [62]:
my_list = [1, 2, 3]

In [63]:
my_list = ["string", 100, 32.2] # can combine datatypes in list

In [64]:
len(my_list) # return length of list

3

In [65]:
my_list = ['one', 'two', 'three']

In [66]:
my_list[0]

'one'

In [67]:
my_list[1:] # slicing of list works like slicing of string

['two', 'three']

In [68]:
another_list = ['four', 3.5]

In [69]:
new_list = my_list + another_list

In [70]:
new_list

['one', 'two', 'three', 'four', 3.5]

In [71]:
new_list[0] = 100 # you can mutate a list, unlike a string

In [72]:
new_list

[100, 'two', 'three', 'four', 3.5]

In [73]:
new_list.append("six") # append a new element to end of list

In [74]:
new_list

[100, 'two', 'three', 'four', 3.5, 'six']

In [75]:
new_list.pop() # remove an elem from the end of list

'six'

In [76]:
new_list

[100, 'two', 'three', 'four', 3.5]

In [77]:
poped_item = new_list.pop() # save the poped item

In [78]:
poped_item

3.5

In [79]:
new_list.pop(0) # remove an elem of specified index

100

In [80]:
new_list

['two', 'three', 'four']

In [81]:
new_list.pop(-2) # reverse index works for lists too

'three'

In [82]:
new_list

['two', 'four']

In [83]:
new_list = ['a', 'e', 'x', 'b', 'c']
num_list = [4, 1, 8, 3]

In [84]:
new_list.sort() # sort the list in-place, so return nothing

In [85]:
new_list

['a', 'b', 'c', 'e', 'x']

In [86]:
num_list.sort()

In [87]:
num_list

[1, 3, 4, 8]

In [90]:
num_list = [23, 535, 213, 4, 6]

In [91]:
num_list.reverse() # reverse the list in-place

In [92]:
num_list

[6, 4, 213, 535, 23]